### PR TITLE
Fixed handling of ambiguous nucleotides, issue #137

### LINKED ISSUE #137: Ambiguous Nucleotide handling question?
How are ambiguous nucleotides (e.g. N) handled? Are the kmer's containing this filtered out?

### PR DESCRIPTION
**Pull request by swamidass**
Fixes #137 by ignoring DNA k-mers that contain sequence other than ACGT.

- [x] Is it mergeable?
- [x] `make test` Did it pass the tests?
- [x] `make coverage` Is the new code covered?
- [x] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [x] Was a spellchecker run on the source code and documentation after
  changes were made?

This change does change the kmers that sourmash will use (throwing out the ambiguous ones), so it will change the signatures of files with ambiguous kmers. Therefore I did increment the version to 1.1.1 as a signal to users that there is a change.

### CODE REVIEW

**Comment by swamidass** [comment]
Also, on my machine, there are 21 failed tests, but this was before the change to the code.

**Comment by swamidass** [comment]
Looks like two small changes needs to be made to the tests (because bad kmers are now being excluded). This is easy to do. If you approve otherwise, I will make the change.

**Comment by ctb** [comment]
A couple of notes --

first, you can leave the checklists as they are, github will make them clickable!

second, we should fix the master branch tests on your system before proceeding. note that they are passing on Travis for both 2.7 and 3.5 (https://travis-ci.org/dib-lab/sourmash) so either we have an error in our install documentation or something wonky is going on on your system.

third, could you add a test (I suggest to `test__minhash.py`, something along the lines of `test_compare_1`) that illustrates the new behavior? It should fail if the patch to `add_sequence` is not present, of course.

thx!

**Comment by swamidass** [comment]
So it looks like the fix works now on Travis. As a separate issue, (even before this fix), the build does not pass on my MacOS laptop. Not sure how to fix that.

**Comment by ctb** [comment]
Thanks, I'll take a look! re your tests, can you post some of the error messages from the failing tests?

**Comment by swamidass** [comment]
See #139. Seems like a lot of save/load errors (which probably totally breaks sourmash here.

**Comment by ctb** [comment]
@swamidass please see https://github.com/swamidass/sourmash/pull/1 -  I realized that the `_forcedna` logic was no longer useful, although it introduces new behavior where valid k-mers are added before the exception is raised. What do you think?

We could always add a `_checkdna` call before the for loop, but then we'd be looking at the entire sequence. *shrug*

**Comment by ctb** [comment]
@betatim @luizirber could you take a quick look to make sure we didn't miss something? I think this is ready for review & merge.

**Comment by swamidass** [comment]
Up to you in that behavior. I merged it in, but that behavior needs to be documented.  It might be better of any not standard nucleotide just gets skipped, rather than raising an exception. That is probably the best behaivior. What do you think?

File: sourmash_lib/test__minhash.py [hunk 1]
```diff
@@ -162,8 +162,9 @@ def test_max_hash_with_limit(track_abundance):
 def test_basic_dna_bad(track_abundance):
     # test behavior on bad DNA
     mh = MinHash(1, 4, track_abundance=track_abundance)
+
     with pytest.raises(ValueError):
-        mh.add_sequence('ATGR')
+        mh.add_sequence('ATGR')
```
Comment (betatim) [inline sourmash_lib/test__minhash.py]: spurious whitespace.

File: sourmash_lib/test__minhash.py [hunk 2]
```diff
@@ -176,14 +177,27 @@ def test_basic_dna_bad_2(track_abundance):
 
 def test_basic_dna_bad_force(track_abundance):
     # test behavior on bad DNA
-    mh = MinHash(1, 4, track_abundance=track_abundance)
+    mh = MinHash(100, 4, track_abundance=track_abundance)
```
Comment (betatim) [inline sourmash_lib/test__minhash.py]: For my curiosity: why the change from `1` to `100`?

File: setup.py [hunk 3]
```diff
@@ -4,7 +4,7 @@
 from setuptools import Extension
 import os
 
-VERSION="1.1"
+VERSION="1.1.1"
```
Comment (betatim) [inline setup.py]: I don't remember, do we change the version number in master prior to a release? If we do I think we should bump it all the way to 2.x.y as that will be the next release and even these changes by themselves would justify going to 2.x.y as the hashes will be incompatible.

File: sourmash_lib/kmer_min_hash.hh [hunk 4]
```diff
@@ -156,29 +167,20 @@ public:
         return aa;
     }
 
-    std::string _checkdna(const char * s, bool force=false) const {
-        std::string seq = s;
-        const size_t seqsize = strlen(s);
+    bool _checkdna(std::string seq) const {
```
Comment (betatim) [inline sourmash_lib/kmer_min_hash.hh]: -> `const std::string seq`

File: sourmash_lib/kmer_min_hash.hh [hunk 5]
```diff
@@ -156,29 +167,20 @@ public:
...
 
-        for (size_t i=0; i < seqsize; ++i) {
+        for (size_t i=0; i < seq.length(); ++i) {
```
Comment (betatim) [inline sourmash_lib/kmer_min_hash.hh]: Would be nice to switch to `for (auto b : seq) {...}` as we touch this code. Up to you.

**Comment by betatim** [comment]
There should be some docs to update for this change. If not we should add to the docs what the new behaviour is.

I've looked at the code/technical part. Whether this is the thing to do or not for science I'll leave to someone else.

**Comment by betatim** [comment]
Did some quick scanning of the docs and found these places that mention ACGTN characters and behaviour:

* http://sourmash.readthedocs.io/en/latest/command-line.html#an-example
* http://sourmash.readthedocs.io/en/latest/api-example.html#consuming-files

**Comment by betatim** [comment]
Can't see the [output of travis](https://travis-ci.org/dib-lab/sourmash/builds/206482938) ATM, I think this is because they use S3 to store them. Will check back later.

**Comment by ctb** [comment]
Skipping non-standard nucleotides is the default behavior from the command
line 'compute' call now. The API does not do so by default.  That seems OK
to me.

**Comment by swamidass** [comment]
What are we waiting for here? 

Given the agreement about incrementing version. The only addition change I could make is changing the increment from 1.1 -> 1.1.1  to either 1.2 or 2.0. 

Please let me know what you suggest, or please merge the change. Thanks.

**Comment by ctb** [comment]
hi @swamidass, @betatim put in a few comments above. Those should be addressed (or discussed if you disagree). I'd be happy to do that, but I'm in traveling and also in the middle of writing a grant, so it may take me another few days.

Is there any particular hurry? You (and others) can use this branch as-is and we're not trying to make a new release any time soon.

**Comment by betatim** [comment]
I am -1 on updating the version. IMO this should happen "as we tag" the next release.

I'd still update the docs to make them more clear on what happens. "ignore non ACGT" is kinda vague, especially if we care at the level of "ignore per kmer, ignore per read". 'here we set ‘force=True’ in add_sequence to ignore non-ACTGN characters' -> 'here we set ‘force=True’ in add_sequence to what-we-actually-do when encountering a non-ACTGN character in a kmer' or some such.

File: sourmash_lib/test__minhash.py (hunk 1 above)
Comment (ctb) [inline sourmash_lib/test__minhash.py]: Fixed in https://github.com/swamidass/sourmash/pull/2

File: sourmash_lib/test__minhash.py (hunk 2 above)
Comment (ctb) [inline sourmash_lib/test__minhash.py]: Added comment in https://github.com/swamidass/sourmash/pull/2 - we want to store multiple hashes.

File: setup.py (hunk 3 above)
Comment (ctb) [inline setup.py]: Updated version to 2.0-alpha in https://github.com/swamidass/sourmash/pull/2

File: sourmash_lib/kmer_min_hash.hh (hunk 4 above)
Comment (ctb) [inline sourmash_lib/kmer_min_hash.hh]: Fixed in https://github.com/swamidass/sourmash/pull/2

File: sourmash_lib/kmer_min_hash.hh (hunk 5 above)
Comment (ctb) [inline sourmash_lib/kmer_min_hash.hh]: Not done yet.

**Comment by ctb** [comment]
I decided to update the version to 2.0-alpha, mostly to indicate that it *is* alpha :)

**Comment by ctb** [comment]
@swamidass if you merge in https://github.com/swamidass/sourmash/pull/2, I think we're ready for merge.

**Comment by ctb** [comment]
I also updated the docs in https://github.com/swamidass/sourmash/pull/2 so that it's clear that k-mers containing non-ACGT are skipped in Python.

**Comment by betatim** [comment]
👍
